pkg/csi/service/vanilla/controller.go

```diff
@@ -884,6 +885,10 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
 		volumeType = prometheus.PrometheusBlockVolumeType
 		node, err := c.nodeMgr.GetNodeByName(ctx, req.NodeId)
 		if err != nil {
+			if err == cnsnode.ErrNodeNotFound {
```

Comment:
If we simply delete Node from Kubernetes Cluster, but does not delete Node from vCenter inventory, this PR will create issue, volume attachment will be gone after this fix, but when user attempts to create a new Pod, we can not attach the volume to the new node, since volume is already attached to the prior node VM.

Reply:
I cannot confirm this scenario. After deleting a node without draining it, the volume attachment was gone and a new pod on another node was able to attach the volume successfully.

Reply:
I think when a node is faulty, the user may need to delete both k8s node and vCenter VM. The user shouldn't be leaving the vCenter VM even after 6 mins ADC timeout. I think it may be Ok to fail the attach because the user did not delete the VM/node properly.
But, when user correctly deletes both k8s node and vCenter VM, I think the unpublish/publish should succeed.

Reply:
@yuga711 I'm a little bit confused now. The method `c.nodeMgr.GetNodeByName` retrieves a `*cnsvsphere.VirtualMachine`, although it is named "node". This is the vCenter VM, isn't it? So if this object cannot be found, can the VM still exist?

Reply:
If the VM is shutdown but not deleted, does `c.nodeMgr.GetNodeByName` return the `ErrNodeNotFound` or something different?

Reply:
by checking the GetNodeByName method above, it looks like checking it though Kubernetes client. Is there a way to actually verify that physical VM still exist through vcenter directly?

Reply:
@MartinWeindel Can you use the below code snippet to get to know if physical VM still exist and then return not found 
```
vm, err := vsphere.GetVirtualMachineByUUID(ctx, nodeUUID, false)
	if err != nil {
		log.Errorf("Couldn't find VM instance with nodeUUID %s, failed to discover with err: %v", nodeUUID, err)
		return err
	}
```

Reply:
@MartinWeindel You can use this : https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/df4513d069fdd723f79d42c4c3408ddc298e0815/pkg/common/cns-lib/node/manager.go#L117 or you can this if you have datacenter object : https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/df4513d069fdd723f79d42c4c3408ddc298e0815/pkg/common/cns-lib/vsphere/datacenter.go#L89

Reply:
Thank you for the information! @chethanv28 

@MartinWeindel could you please add the logic to verify VM exist from vSphere?

Reply:
Yes I will work on it asap.

Reply:
Using `GetVirtualMachineByUUID` is not possible, as the K8s node object containing the providerID (which encodes the UUID) is already gone. Instead I had to introduce a method `GetVirtualMachineByDNSName` to lookup the vCenter VM, as the nodeID is always used as DNS name as far as I know. This is a similar approach as used in the cloud-provider-vsphere (see https://github.com/kubernetes/cloud-provider-vsphere/blob/master/pkg/cloudprovider/vsphere/nodemanager.go#L111).
I have tried it out in my test environment and the volume attachment is only deleted if the VM is gone.

Reply:
@divyenpatel You mentioned GetVirtualMachineByDNSName might not work in some cases. Could you put details here? Is there any way to solve it?
@xing-yang

Reply:
IIRC, Divyen said the GetVirtualMachineByDNSName() can potentially return a different VM. 
Do we have a vCenter API that gives the VM where the volume is attached?

cc @gnufied

Reply:
can we make assumption that if K8s Node object should not be deleted until VM is deleted from vcenter inventory. In what case user wants to delete k8s Node object without deleting it from vCenter inventory? 
User might break this assumption, but it should be treated as user error. User might do other things to break the system, such as deleting VolumeAttachement objects while volumes are still attached.

Reply:
@divyenpatel 
For intree driver, we return success if the node is not found in nodemanager cache. 
https://github.com/kubernetes/kubernetes/blob/770a9504ea3fe38a75c51f036187b5dea124ebab/staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go#L997
Is it ok if we do something similar for CSI?